File: src/Render/Camera.cpp
#include "Render/Camera.h"
#include <glm/gtc/matrix_transform.hpp>
#include <algorithm>
#include <cmath>

namespace mc {

static constexpr float DEG2RAD = 3.14159265358979f / 180.0f;

glm::vec3 Camera::front() const {
    float cy = std::cos(yaw * DEG2RAD), sy = std::sin(yaw * DEG2RAD);
    float cp = std::cos(pitch * DEG2RAD), sp = std::sin(pitch * DEG2RAD);
    return glm::normalize(glm::vec3(cy * cp, sp, sy * cp));
}

glm::vec3 Camera::forwardXZ() const {
    float cy = std::cos(yaw * DEG2RAD), sy = std::sin(yaw * DEG2RAD);
    return glm::normalize(glm::vec3(cy, 0.0f, sy));
}

glm::vec3 Camera::right() const {
    return glm::normalize(glm::cross(front(), glm::vec3(0, 1, 0)));
}

glm::vec3 Camera::up() const {
    return glm::normalize(glm::cross(right(), front()));
}

glm::mat4 Camera::view() const {
    return glm::lookAt(position, position + front(), glm::vec3(0, 1, 0));
}

glm::mat4 Camera::projection() const {
    return glm::perspective(glm::radians(fov), aspect, nearPlane, farPlane);
}

void Camera::addMouseDelta(float dx, float dy, float sensitivity) {
    yaw += dx * sensitivity;
    pitch -= dy * sensitivity;
    pitch = std::clamp(pitch, -89.5f, 89.5f);
}

void Camera::updateFrustum() {
    glm::mat4 m = projection() * view();
    // Gribb/Hartmann plane extraction.
    for (int i = 0; i < 3; ++i) {
        planes_[i * 2 + 0] = glm::vec4(m[0][3] + m[0][i], m[1][3] + m[1][i],
                                       m[2][3] + m[2][i], m[3][3] + m[3][i]);
        planes_[i * 2 + 1] = glm::vec4(m[0][3] - m[0][i], m[1][3] - m[1][i],
                                       m[2][3] - m[2][i], m[3][3] - m[3][i]);
    }
    for (auto& p : planes_) {
        float len = glm::length(glm::vec3(p));
        if (len > 0) p /= len;
    }
}

bool Camera::aabbVisible(const glm::vec3& mn, const glm::vec3& mx) const {
    for (const auto& p : planes_) {
        glm::vec3 n(p);
        // Pick the positive vertex (farthest along plane normal).
        glm::vec3 pv(n.x >= 0 ? mx.x : mn.x,
                     n.y >= 0 ? mx.y : mn.y,
                     n.z >= 0 ? mx.z : mn.z);
        if (glm::dot(n, pv) + p.w < 0) return false;
    }
    return true;
}

} // namespace mc

File: minicraft-java/src/main/java/com/minicraft/render/Camera.java
package com.minicraft.render;

import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.joml.Vector4f;

/** View/projection provider plus a frustum for chunk culling. */
public class Camera {
    public final Vector3f position = new Vector3f(0, 80, 0);
    public float yaw = -90f, pitch = 0f;
    public float fov = 70f, aspect = 16f / 9f;
    public float nearPlane = 0.05f, farPlane = 1000f;

    private final Vector4f[] planes = new Vector4f[6];

    public Camera() {
        for (int i = 0; i < 6; i++) planes[i] = new Vector4f();
    }

    public Vector3f front() {
        double cy = Math.cos(Math.toRadians(yaw)), sy = Math.sin(Math.toRadians(yaw));
        double cp = Math.cos(Math.toRadians(pitch)), sp = Math.sin(Math.toRadians(pitch));
        return new Vector3f((float)(cy * cp), (float) sp, (float)(sy * cp)).normalize();
    }

    public Vector3f forwardXZ() {
        double cy = Math.cos(Math.toRadians(yaw)), sy = Math.sin(Math.toRadians(yaw));
        return new Vector3f((float) cy, 0, (float) sy).normalize();
    }

    public void addMouseDelta(float dx, float dy, float sensitivity) {
        yaw += dx * sensitivity;
        pitch -= dy * sensitivity;
        if (pitch > 89.5f) pitch = 89.5f;
        if (pitch < -89.5f) pitch = -89.5f;
    }

    public Matrix4f view() {
        Vector3f f = front();
        Vector3f center = new Vector3f(position).add(f);
        return new Matrix4f().lookAt(position, center, new Vector3f(0, 1, 0));
    }

    public Matrix4f projection() {
        return new Matrix4f().perspective((float) Math.toRadians(fov), aspect, nearPlane, farPlane);
    }

    public void updateFrustum() {
        Matrix4f m = projection().mul(view());
        // Gribb/Hartmann extraction.
        set(0, m.m03() + m.m00(), m.m13() + m.m10(), m.m23() + m.m20(), m.m33() + m.m30());
        set(1, m.m03() - m.m00(), m.m13() - m.m10(), m.m23() - m.m20(), m.m33() - m.m30());
        set(2, m.m03() + m.m01(), m.m13() + m.m11(), m.m23() + m.m21(), m.m33() + m.m31());
        set(3, m.m03() - m.m01(), m.m13() - m.m11(), m.m23() - m.m21(), m.m33() - m.m31());
        set(4, m.m03() + m.m02(), m.m13() + m.m12(), m.m23() + m.m22(), m.m33() + m.m32());
        set(5, m.m03() - m.m02(), m.m13() - m.m12(), m.m23() - m.m22(), m.m33() - m.m32());
        for (Vector4f p : planes) {
            float len = (float) Math.sqrt(p.x * p.x + p.y * p.y + p.z * p.z);
            if (len > 0) { p.x /= len; p.y /= len; p.z /= len; p.w /= len; }
        }
    }

    private void set(int i, float a, float b, float c, float d) { planes[i].set(a, b, c, d); }

    public boolean aabbVisible(float minX, float minY, float minZ, float maxX, float maxY, float maxZ) {
        for (Vector4f p : planes) {
            float px = p.x >= 0 ? maxX : minX;
            float py = p.y >= 0 ? maxY : minY;
            float pz = p.z >= 0 ? maxZ : minZ;
            if (p.x * px + p.y * py + p.z * pz + p.w < 0) return false;
        }
        return true;
    }
}
